### DISCUSSION TITLE
Trying to solve a lifecycle problem where A depends on B, but B gets disposed too early

### DISCUSSION THREAD

**Discussion by derolf** [rrousselGit/riverpod#2675] [Q&A]
Consider this example

```
import "package:riverpod/riverpod.dart";

final loggingProvider =
    AutoDisposeNotifierProvider<Logger, List<String>>(() => Logger());

class Logger extends AutoDisposeNotifier<List<String>> {
  @override
  List<String> build() {
    state = [];
    log("Logger started");
    ref.onDispose(
      () {
        log("Logger disposed");
        _isDisposed = true;
        _printLogs();
      },
    );
    return state;
  }

  void log(String msg) {
    assert(!_isDisposed);
    state = [...state, msg];
  }

  void _printLogs() => state.forEach(print);

  var _isDisposed = false;
}

final valueProvider = StateProvider.autoDispose(
  (ref) {
    final logger = ref.watch(loggingProvider.notifier);

    // -> Providers are not allowed to modify other providers during their initialization.
    try {
      logger.log(
        "valueProvider created",
      );
    } catch (e) {
      print("1 $e");
    }

    // -> Failed assertion: line 22 pos 12: '!_isDisposed': is not true.
    Future(
      () {
        try {
          logger.log(
            "valueProvider created",
          );
        } catch (e) {
          print("2 $e");
        }
      },
    );

    ref.onDispose(
      () {
        logger.log(
          "valueProvider disposed",
        );
      },
    );

    return 0;
  },
);

//  fvm flutter pub run --enable-asserts test/lifecycle_demo.dart

void main() {
  final container = ProviderContainer();
  container.read(valueProvider.notifier).state = 1;
  container.dispose();
}
```

The output is

```
1 'package:riverpod/src/framework/element.dart': Failed assertion: line 478 pos 11: '_debugCurrentlyBuildingElement == null ||
              _debugCurrentlyBuildingElement == this': Providers are not allowed to modify other providers during their initialization.

The provider AutoDisposeStateProvider<int>#616a4 modified AutoDisposeNotifierProviderImpl<Logger, List<String>>#10fc9 while building.

Logger started
valueProvider created
valueProvider disposed
Logger disposed
2 'file:///...lifecycle_demo.dart': Failed assertion: line 22 pos 12: '!_isDisposed': is not true.
```

How would you solve the problem?

The underlying problem is: If we use futures to defer events, how can we ensure that the underlying 'controllers' aren't getting disposed before we can act on the deferred event.

**Comment by rrousselGit** [comment]
Why is logger a notifier in your case? I'm not sure you need to listen to logs using ref.watch/listen, do you?

If the logger wasn't a notifier but rather a normal class, your code would likely work.

**Reply by derolf** [reply]
There might be a widget in the app that watches the logs and displays them.

**Reply by rrousselGit** [reply]
The error is fully expected then.

**Comment by rrousselGit** [comment]
> The underlying problem is: If we use futures to defer events, how can we ensure that the underlying 'controllers' aren't getting disposed before we can act on the deferred event.

I don't understand this question.

**Reply by derolf** [reply]
If you use a Future and then call the controller, it might already got disposed.

**Reply by rrousselGit** [reply]
Right, because your provider/notifier got destroyed before the future completed.
You should cancel the future.

**Reply by derolf** [reply]
Would getting some sort of keepAlive "handle" on the underlying provider help?

**Reply by rrousselGit** [reply]
Generally yes. But in your case you're calling `ProviderContainer.dispose` itself. That's not going to help here. The provider is always going to get destroyed immediately, even if it wasn't autoDispose at all

**Reply by derolf** [reply]
That’s just for the sake of “test”.

**Reply by rrousselGit** [reply]
Right, but then your test is not representative of your actual situation.
In that case, a watch/listen somewhere to keep the provider alive should do it, yes.
